framework/calico_framework.py

```diff
@@ -0,0 +1,661 @@
+                        continue
+
+
+                if offer.num_tasks_launchable < len(tasks):
+                    raise NotEnoughResources("Need a larger offer to meet slave-id specifications")
+
+                # We can fit the tasks for this slave on this offer.
+                for task in tasks:
+                    _log.debug("\t\tPlanning %s for %s", task, offer)
+                    task.slave_id = offer.offer.slave_id.value
+                    offer.reserved_by = self
+
+            # Check that all anywhere_tasks were assigned
+            for task in anywhere_tasks:
+                if task.slave_id == None:
+                    raise NotEnoughResources("Matched all slave-specific tasks, but not enough remaining resources to launch "
+                                             "the remaining %d anywhere-tasks" % len(anywhere_tasks))
+        except NotEnoughResources as e:
+            # Rollback all reservations
+            _log.info("\t\tNot Launching: %s" % e)
+            for task in self.tasks:
+                task.slave_id = None
+
+            for offer in offers:
+                if offer.reserved_by == self:
+                    offer.reserved_by = None
+            return False
+        else:
+            return True
+
+    def print_report(self):
+        """
+        Print this test's results.
+        """
+        print "|--- %s ---|" % self.name
+        failed_tasks = self.tasks_with_state(BAD_TASK_STATES)
+        if self.killed:
+            print "Test Status: KILLED"
+        elif failed_tasks:
+            print "Test Status: FAIL"
+        elif self.tasks_with_state([None]) == self.tasks:
+            print "Test Status: UNLAUNCHED"
+        else:
+            print "Test Status: PASS"
+
+        print "Reason: ", self.additional_info
+        print "Task Statuses:"
+        for task in self.tasks:
+            if task.state == None:
+                state = "UNLAUNCHED"
+            else:
+                state = mesos_pb2.TaskState.Name(task.state)
+            print "\t%s: %s" % (task, state)
+            if type(task) == PingTask:
+                print "\t\tTest Results: %s" % task.ping_status_data
+
+        print "-----------------------------------"
+
+    def tasks_with_state(self, states):
+        return [task for task in self.tasks if task.state in states]
+
+    @property
+    def running(self):
+        if self.killed:
+            return False
+        else:
+            return self.tasks_with_state([mesos_pb2.TASK_RUNNING, mesos_pb2.TASK_STAGING]) != []
+
+    def launch(self, offers):
+        """
+        Launches the test on the available offers.
+        This assumes the test has already reserved offers via self.can_run_on, and
+        assumes each task has already picked a task.slave_id.
+        """
+        # Keep track of offers that were accepted so we can remove them from the Scheduler
+        accepted_offers = []
+        for offer in offers:
```

Comment:
I dunno, maybe I'm making a big deal out of nothing.  I don't think any tests or clusters this will run on will be larger than a few dozen of tasks or offers...